File: 프로그래머스/2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.py
def dfs(o, n, k, ls, rs):
    if n == k:
        nl = [o[i] for i in ls]
        nl.sort()
        rs.append("".join(nl))
        return

    for i in range(len(o)):
        if not ls or ls[-1] < i:
            ls.append(i)
            dfs(o, n + 1, k, ls, rs)
            ls.pop()


def include(a, b):
    return all([s in b for s in list(a)])


def solution(orders, course):
    result = []

    for size in course:
        size_set = set()
        for o in orders:
            rs = []
            dfs(o, 0, size, [], rs)
            for r in rs:
                size_set.add(r)

        dic = {}
        for r in size_set:
            cnt = 0
            for o in orders:
                if include(r, o):
                    cnt += 1
            dic[r] = cnt

        if not dic:
            continue
        # print(dic)
        items = dic.items()
        maxv = max(dic.values())
        maxv = 2 if maxv < 2 else maxv

        items = [k for k, v in items if v == maxv]
        for it in items:
            result.append("".join(sorted(it)))

    result.sort()
    return result

File: 프로그래머스/2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.java
import java.util.*;
import java.util.stream.*;


class Solution {
    List<String> result = new ArrayList();
    LinkedList<Integer> ls;
    Map<String, Integer> map;
    
    void Dfs(String[] n_order, int k){
        if (ls.size() == k){
            int[] ints = ls.stream().mapToInt(i -> n_order[i].charAt(0)).toArray();
            char[] chars = new char[ints.length];
            for (int i = 0; i < ints.length; i++){
                chars[i] = (char)ints[i];
            }
            
            String sub = String.valueOf(chars);
            int cnt = map.getOrDefault(sub, 0) + 1;
            map.put(sub, cnt);
            return;
        }
        
        for (int i = 0; i < n_order.length; i++){
            if (ls.size() == 0 || ls.peekLast() < i){
                ls.addLast(i);
                Dfs(n_order, k);
                ls.pollLast();
            }
        }
    }
    
    
    
    public String[] solution(String[] orders, int[] course) {
        
        for(int size : course){
            map = new HashMap();
            for(String order : orders){
                String[] n_order = Arrays.stream(order.split("")).sorted().toArray(String[]::new);
                ls = new LinkedList();
                Dfs(n_order ,size);                
            }
            
            if (map.isEmpty()) continue;
            
            int maxv = map.values().stream().mapToInt(i->(int)i).max().orElse(0);
            final int max = Math.max(maxv, 2);
            List<String> list = map.entrySet().stream()
                .filter(entry -> entry.getValue() >= max)
                .map(entry -> entry.getKey())
                .collect(Collectors.toList());
            
            for(String s : list){
                result.add(s);
            }
        }
        
        return result.stream().sorted().toArray(String[]::new);
    }
}